packages/synapse-core/src/utils/cert.ts

```diff
@@ -0,0 +1,150 @@
+import { signTypedData } from 'viem/actions'
+import { randU256 } from '../utils/rand.ts'
+
+export type Endorsement = {
+  /**
+   * Unique nonce to suport nonce based revocation.
+   */
+  nonce: bigint
+  /**
+   * This certificate becomes invalid after `notAfter` timestamp.
+   */
+  notAfter: bigint
+}
+
+export type SignedEndorsement = Endorsement & {
+  signature: Hex
+}
+
+export const EIP712Endorsement = {
+  Endorsement: [
+    { name: 'nonce', type: 'uint64' },
+    { name: 'notAfter', type: 'uint64' },
+    { name: 'providerId', type: 'uint256' },
+  ],
+} as const
+
+export type TypedEn = TypedDataToPrimitiveTypes<typeof EIP712Endorsement>['Endorsement']
+
+export type SignCertOptions = {
+  nonce?: bigint // uint64
+  notAfter: bigint // uint64
+  providerId: bigint
+} /**
+ * Signs a certificate that a provider is super good enough.
+ * @param client - The client to use to sign the message
+ * @param options - nonce (randomised if null), not after and who to sign it for
+ * @returns encoded certificate data abiEncodePacked([nonce, notAfter, signature]), the provider id is implicit by where it will get placed in registry.
+ */
+export async function signEndorsement(client: Client<Transport, Chain, Account>, options: SignCertOptions) {
+  const nonce = (options.nonce ?? randU256()) & 0xffffffffffffffffn
+  const signature = await signTypedData(client, {
+    account: client.account,
+    domain: {
+      name: 'Storage Endorsement',
+      version: '1',
+      chainId: client.chain.id,
+    },
+    types: EIP712Endorsement,
+    primaryType: 'Endorsement',
+    message: {
+      nonce: nonce,
+      notAfter: options.notAfter,
+      providerId: options.providerId,
+    },
+  })
+
+  const encodedNonce = numberToHex(nonce, { size: 8 })
+  const encodedNotAfter = numberToHex(options.notAfter, { size: 8 })
+
+  return concat([encodedNonce, encodedNotAfter, signature])
+}
+
+export async function decodeEndorsement(
+  providerId: bigint,
+  chainId: number | bigint | undefined,
+  hexData: Hex
+): Promise<{
+  address: Address | null
+  endorsement: SignedEndorsement
+}> {
+  if (hexData.length !== 164) {
+    return {
+      address: null,
+      endorsement: {
+        nonce: 0n,
+        notAfter: 0n,
+        signature: '0x',
+      },
+    }
+  }
+  const data = hexToBytes(hexData)
+  const endorsement: SignedEndorsement = {
+    nonce: bytesToBigInt(data.slice(0, 8)),
+    notAfter: bytesToBigInt(data.slice(8, 16)),
+    signature: bytesToHex(data.slice(16)),
+  }
+  const address = await recoverTypedDataAddress({
+    domain: {
+      name: 'Storage Endorsement',
+      version: '1',
+      chainId,
+    },
+    types: EIP712Endorsement,
+    primaryType: 'Endorsement',
+    message: {
+      nonce: endorsement.nonce,
+      notAfter: endorsement.notAfter,
+      providerId: providerId,
+    },
+    signature: endorsement.signature,
+  }).catch((reason) => {
+    console.warn('Failed to recover product endorsement:', reason)
```

Comment:
since we're throwing all capabilities at this function this could get noisy, perhaps we not bother

```suggestion
```